092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,300 @@
+
+##### KubernetesEventsNotifier 
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `KubernetesEventsNotifier` will be based of the `SelfHealingNotifier` class.
+The `approve/ignore/delay` process for `KubernetesEventNotifier` will be similar to that of the `SelfHealingNotifier` class. 
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable then they will be ignored.
+For effective alerts and monitoring, it makes use of Kubernetes events to signal to users what operations Cruise Control is performing automatically.
+The `KubernetesEventsNotifier` will override the `alert()` method of the `SelfHealingNotifier` and will trigger and publish events whenever an anomaly is detected.
+These events would publish information regarding the anomaly like the anomalyID, the type of anomaly which was detected etc.
+The events would help the user to understand the anomalies that are present, when the anomalies appeared in their cluster and also would help them know what is going on in their clusters.
+
+The event published would look like this:
+
+```yaml
+- action: <action>
+  apiVersion: v1
+  eventTime: "2024-12-19T08:06:21.314140Z"
+  firstTimestamp: null
+  involvedObject:
+    kind: Kafka
+    name: <cluster-name>
+    namespace: <project-namespace>
+  kind: Event
+  lastTimestamp: null
+  message: Cruise Control is fixing the detected anomaly
+  metadata:
+    creationTimestamp: "2024-12-19T08:06:21Z"
+    generateName: cruise-control-event
+    name: my-cluster-34c41626-0add-47a3-9b79-d6ee29d5d81f  # my-cluster-<Anomaly ID>
+    namespace: myproject
+    resourceVersion: "101070"
+    uid: 80892c7d-e3a1-44ad-84dc-2d88e15a7ff0
+  reason: Anomaly was detected in the cluster GOAL_VIOLATION # Detected Anomaly type
+  related: 
+    kind: Pod
+    name: <cruise-control-pod-name>
+    namespace: <project-namespace>
+  reportingComponent: cruise-control
+  reportingInstance: cruise-control
+  source: {}
+  type: Normal
+```
+
+### Strimzi Operations while self-healing is running/fixing an anomaly 
+
+#### What happens when self-healing is running/fixing an anomaly and brokers are rolled
+
+When an anomaly is detected, Cruise Control may try to fix/mitigate it by moving the partition replicas across the brokers in the Kafka cluster.
+If, during the process of moving partitions replicas among the brokers in the Kafka cluster, some broker that was being used gets deleted or rolled, then Cruise Control would finish self-healing process and log that `self-healing finished successful with error`.
```

Comment:
What does `finish` mean in this context? It sounds like it will move a partition replica to a black hole left after a removed broker which would be pretty nasty. I assume it will `stop` the process with an error instead?

Reply:
so the process is stopped but I don't think the there are any reverts, so its more like incomplete movement. The upstream CC logs still states finished succesful with error. Maybe we shoud get this error message fixed upstream